otter/convergence/gathering.py

```diff
@@ -112,8 +116,71 @@ def group_id(s):
                             filter(server_predicate),
                             filter(has_group_id))
 
-    eff = get_all_server_details(changes_since)
-    return eff.on(servers_apply)
+    return get_all_server_details(changes_since).on(servers_apply)
+
+
+def merge_servers(first, second):
+    """
+    Return servers got by merging the two sets of servers where second
+    takes precedence over first
+
+    :param list first: List of server dicts
+    :param list second: List of server dicts
+    :return: List of merged server dicts
+    """
+    def sdict(servers):
+        return {s['id']: s for s in servers}
+
+    return merge(sdict(first), sdict(second)).values()
+
+
+def server_of_group(group_id, server):
+    """
+    Return True if server belongs to group_id. False otherwise
+    """
+    return group_id_from_metadata(server.get('metadata', {})) == group_id
+
+
+def group_servers(group_id, grouped_servers):
+    """
+    Return servers of a group defaulting to empty if not found
+    """
+    return grouped_servers.get(group_id, [])
+
+
+@do
+def get_scaling_group_servers(tenant_id, group_id, now,
+                              cache_class=CassScalingGroupServersCache,
+                              all_as_servers=get_all_scaling_group_servers,
+                              all_servers=get_all_server_details):
+    """
+    Get a group's servers taken from cache if it exists. Updates cache
+    if it is empty from newly fetched servers
+    # NOTE: This function takes tenant_id even though the whole effect is
+    # scoped on the tenant because cache calls require tenant_id. Should
+    # they also not take tenant_id and work on the scope?
+
+    :return: Servers as list of dicts
+    :rtype: Effect
+    """
+
+    cache = cache_class(tenant_id, group_id)
+    cached_servers, last_update = yield cache.get_servers()
+    if last_update is None:
+        servers = (yield all_as_servers()).get(group_id, [])
+        yield cache.insert_servers(now, servers, True)
+    elif now - last_update >= timedelta(days=30):
+        last_update = now - timedelta(days=30)
+        changes, current = yield parallel([
+            all_as_servers(last_update).on(partial(group_servers, group_id)),
+            all_as_servers().on(partial(group_servers, group_id))])
+        servers = merge_servers(changes, current)
+        yield cache.insert_servers(now, servers, True)
+    else:
+        changes = yield all_servers(last_update)
+        servers = merge_servers(cached_servers, changes)
+        servers = list(filter(partial(server_of_group, group_id), servers))
```

Comment:
I can't really remember what the conclusion was about why we're not updating the servers cache in this case either, so perhaps you could add a code comment here explaining why?